removed normalized_avg_processors and created documentation on how to add it back (#1981)
ubccr/xdmod: docs/howto-normalized-avg-processors.md

Reviewer: I think there should be a description at the top of this file explaining why the statistic was removed.

Reviewer 2: Definitely

> The "Job Size: Normalized" statistic can be restored if you modify/add the following JSON and SQL files to your installation of Open XDMoD. The files are located in `/etc/xdmod` unless you used a different prefix when installing Open XDMoD (e.g., `/opt/xdmod-{{ page.sw_version }}/etc`).

Reviewer 2: ```suggestion
The "Job Size: Normalized" statistic in the Jobs realm was removed from the default list
of statistics in version 11.5.0 of Open XDMoD. This was done because the values displayed in XDMoD were incorrect if the resource size changed during the plotted time period. The statistic computation also did not produce a pysically meaningful numerical values except when grouping by or filtering by resource.

The recommended way to report on this type of statistic in XDMoD 11.5+ is to use the Data Analytics Framework to compute the normalization factor. However, the "Job Size: Normalized" statistic can be restored to the portal by making the following changes to  JSON and SQL files to your installation of Open XDMoD. For an RPM-based install the files are located in `/etc/xdmod` for a source code install they are located in the`/etc` directory under the installation location.
```

> The "Job Size: Normalized" statistic in the Jobs realm was removed from the default list of statistics in version 11.5.0 of Open XDMoD. The statistic was removed because the values displayed in XDMoD were incorrect if the resource size changed during the plotted time period. The statistic computation also did not produce a physically meaningful numerical values except when grouping by or filtering by resource.

Reviewer: ```suggestion
The "Job Size: Normalized" statistic in the Jobs realm was removed from the default list of statistics in version 11.5.0 of Open XDMoD because the values of the statistic were incorrect if the resource size changed during the time period of the dataset. The values of the statistic were also not physically meaningful except when grouping or filtering by a single resource.
```

> The recommended way to report on this type of statistic in Open XDMoD 11.5+ is to use the Data Analytics Framework to compute the normalization factor. However, the "Job Size: Normalized" statistic can be restored to the  Open XDMoD portal by making the following changes to  configuration files in your installation of Open XDMoD. For an RPM-based install, these files are located in `/etc/xdmod`. For a source code install they are located in the`/etc` directory under the installation location.

Reviewer: ```suggestion
The recommended way to report on this type of statistic in Open XDMoD 11.5+ is to use the [Data Analytics Framework](data-analytics-framework.md) to fetch data from your Open XDMoD portal and compute the normalization factor. However, the "Job Size: Normalized" statistic can also be restored to your Open XDMoD portal by making the following changes to configuration files in your installation of Open XDMoD.
```

        "name": "Job Size: Normalized",
        "precision": 1,
        "timeseries_formula": {
            "$include": "datawarehouse.d/include/Jobs-norm-avg-proc-time.sql"
        },
        "unit": "% of Total Cores"
    },

```

2. Add the following to `datawarehouse.d/ref/Gateways-statistics.json`:
```json
 "normalized_avg_processors": {
        "$overwrite": {
            "description_html": "The percentage average size ${ORGANIZATION_NAME} ${REALM_NAME} job over total machine cores.<br><i>Normalized Job Size: </i>The percentage total number of processor cores used by a (parallel) job over the total number of cores on the machine.",
            "name": "Job Size: Normalized via ${REALM_NAME}"

>         },
>         "$ref-with-overwrite": "datawarehouse.d/ref/Jobs-statistics.json#/normalized_avg_processors"
>     },
> ```
> 
> 3. Create the file `datawarehouse.d/include/Jobs-norm-avg-proc-time.sql`
> ```sql
> 100.0 *
> COALESCE(
>     SUM(agg.processor_count * agg.running_job_count)
>     /
>     SUM(agg.running_job_count)
>     /
>     (
>         SELECT
>             SUM(rrf.cpu_processor_count)
>         FROM
>             modw.resourcespecs rrf
>         WHERE
>             FIND_IN_SET(
>                 rrf.resource_id,
>                 GROUP_CONCAT(distinct agg.task_resource_id)
>             ) <> 0
>             AND ${AGGREGATION_UNIT}_end_ts >= rrf.start_date_ts
>             AND (
>                 rrf.end_date_ts IS NULL
>                 OR ${AGGREGATION_UNIT}_end_ts <= rrf.end_date_ts
>             )
>     ),
> 0)
> 
> ```
> 
> 4. Create the file `datawarehouse.d/include/Jobs-norm-avg-proc-agg.sql`:
> ```sql
> 100.0 *
> COALESCE(
>     SUM(agg.processor_count * CASE ${DATE_TABLE_ID_FIELD} WHEN ${MIN_DATE_ID} THEN agg.running_job_count ELSE agg.started_job_count END)
>     /
>     SUM(CASE ${DATE_TABLE_ID_FIELD} WHEN ${MIN_DATE_ID} THEN agg.running_job_count ELSE agg.started_job_count END)
>     /
>     (
>         SELECT
>             SUM(rrf.cpu_processor_count)
>         FROM
>             modw.resourcespecs rrf
>         WHERE
>             FIND_IN_SET(
>                 rrf.resource_id,
>                 GROUP_CONCAT(distinct agg.task_resource_id)
>             ) <> 0
>             AND ${AGGREGATION_UNIT}_end_ts >= rrf.start_date_ts
>             AND (
>                 rrf.end_date_ts IS NULL
>                 OR ${AGGREGATION_UNIT}_end_ts <= rrf.end_date_ts
>             )
>     ),
> 0)
> 
> ```

Reviewer: ```suggestion
1. Add the following statistic to `datawarehouse.d/ref/Jobs-statistics.json`:
    ```json
    "normalized_avg_processors": {
        "aggregate_formula": {
            "$include": "datawarehouse.d/include/Jobs-norm-avg-proc-agg.sql"
        },
        "description_html": "The percentage average size ${ORGANIZATION_NAME} job divided by the total number of cores in the resource where the job ran. The job normalization calculation assumes that the resource size is constant. This statistic should not be used with a time range where the resource size changes, because the statistic will be incorrect.<br><i>Normalized Job Size: </i>The ratio of the total number of processor cores used by a (parallel) job over the total number of cores on the resource.",
        "name": "Job Size: Normalized",
        "precision": 1,
        "timeseries_formula": {
            "$include": "datawarehouse.d/include/Jobs-norm-avg-proc-time.sql"
        },
        "unit": "% of Total Cores"
    }
    ```
2. Add the following statistic to `datawarehouse.d/ref/Gateways-statistics.json`:
    ```json
    "normalized_avg_processors": {
        "$overwrite": {
            "description_html": "The percentage average size ${ORGANIZATION_NAME} ${REALM_NAME} job over total machine cores.<br><i>Normalized Job Size: </i>The percentage total number of processor cores used by a (parallel) job over the total number of cores on the machine.",
            "name": "Job Size: Normalized via ${REALM_NAME}"
        },
        "$ref-with-overwrite": "datawarehouse.d/ref/Jobs-statistics.json#/normalized_avg_processors"
    }
    ```
3. Create the file `datawarehouse.d/include/Jobs-norm-avg-proc-time.sql`:
    ```sql
    100.0 *
    COALESCE(
        SUM(agg.processor_count * agg.running_job_count)
        /
        SUM(agg.running_job_count)
        /
        (
            SELECT
                SUM(rrf.cpu_processor_count)
            FROM
                modw.resourcespecs rrf
            WHERE
                FIND_IN_SET(
                    rrf.resource_id,
                    GROUP_CONCAT(distinct agg.task_resource_id)
                ) <> 0
                AND ${AGGREGATION_UNIT}_end_ts >= rrf.start_date_ts
                AND (
                    rrf.end_date_ts IS NULL
                    OR ${AGGREGATION_UNIT}_end_ts <= rrf.end_date_ts
                )
        ),
    0)
    ```
4. Create the file `datawarehouse.d/include/Jobs-norm-avg-proc-agg.sql`:
    ```sql
    100.0 *
    COALESCE(
        SUM(agg.processor_count * CASE ${DATE_TABLE_ID_FIELD} WHEN ${MIN_DATE_ID} THEN agg.running_job_count ELSE agg.started_job_count END)
        /
        SUM(CASE ${DATE_TABLE_ID_FIELD} WHEN ${MIN_DATE_ID} THEN agg.running_job_count ELSE agg.started_job_count END)
        /
        (
            SELECT
                SUM(rrf.cpu_processor_count)
            FROM
                modw.resourcespecs rrf
            WHERE
                FIND_IN_SET(
                    rrf.resource_id,
                    GROUP_CONCAT(distinct agg.task_resource_id)
                ) <> 0
                AND ${AGGREGATION_UNIT}_end_ts >= rrf.start_date_ts
                AND (
                    rrf.end_date_ts IS NULL
                    OR ${AGGREGATION_UNIT}_end_ts <= rrf.end_date_ts
                )
        ),
    0)
    ```
```
Indenting so the numbered list is formatted properly and removing extra blank lines.

> The Normalized Avg Processor statistic is available if you add the following JSON and SQL Files to the XDMoD project.

Reviewer: ```suggestion
The "Job Size: Normalized" statistic can be restored if you add the following JSON and SQL files to your installation of Open XDMoD. The files are located in `/etc/xdmod` unless you used a different prefix when installing Open XDMoD (e.g., `/opt/xdmod-{{ page.sw_version }}/etc`).
```

> 1. Add the follwing to:
> - configuration/datawarehouse.d/ref/Job-Statistics.json

Reviewer: ```suggestion
Add the following to `datawarehouse.d/ref/Jobs-statistics.json`:
```

```json
    "normalized_avg_processors": {
        "aggregate_formula": {
            "$include": "datawarehouse.d/include/Jobs-norm-avg-proc-agg.sql"
        },
        "description_html": "The percentage average size ${ORGANIZATION_NAME} job divided by the total number of cores in the resource where the job ran. The job normalization calculation assumes that the resource size is constant. This statistic should not be used with a time range where the resource size changes, because the statistic will be incorrect.<br><i>Normalized Job Size: </i>The ratio of the total number of processor cores used by a (parallel) job over the total number of cores on the resource.",
        "name": "Job Size: Normalized",
        "precision": 1,
        "timeseries_formula": {
            "$include": "datawarehouse.d/include/Jobs-norm-avg-proc-time.sql"
        },
        "unit": "% of Total Cores"
    },

```

> 2. Add the following to:
> - configuration/datawarehouse.d/ref/Gateways-Statistics.json

Reviewer: ```suggestion
2. Add the following to `datawarehouse.d/ref/Gateways-statistics.json`:
```

```json
 "normalized_avg_processors": {
        "$overwrite": {
            "description_html": "The percentage average size ${ORGANIZATION_NAME} ${REALM_NAME} job over total machine cores.<br><i>Normalized Job Size: </i>The percentage total number of processor cores used by a (parallel) job over the total number of cores on the machine.",
            "name": "Job Size: Normalized via ${REALM_NAME}"
        },
        "$ref-with-overwrite": "datawarehouse.d/ref/Jobs-statistics.json#/normalized_avg_processors"
    },
```

> 3. Create a **jobs-norm-avg-proc-time.sql** in the following:
> - configuration/datawarehouse.d/include/jobs-norm-avg-proc-time.sql

Reviewer: ```suggestion
3. Create the file `datawarehouse.d/include/Jobs-norm-avg-proc-time.sql`:
```

```sql
100.0 *
COALESCE(
    SUM(agg.processor_count * agg.running_job_count)
    /
    SUM(agg.running_job_count)
    /
    (
        SELECT
            SUM(rrf.cpu_processor_count)
        FROM
            modw.resourcespecs rrf
        WHERE
            FIND_IN_SET(
                rrf.resource_id,
                GROUP_CONCAT(distinct agg.task_resource_id)
            ) <> 0
            AND ${AGGREGATION_UNIT}_end_ts >= rrf.start_date_ts
            AND (
                rrf.end_date_ts IS NULL
                OR ${AGGREGATION_UNIT}_end_ts <= rrf.end_date_ts
            )
    ),
0)

```

> 4. create a **jobs-norm-avg-proc-agg.sql** at the follwing:
> - configuration/datawarehouse.d/include/jobs-norm-avg-proc-agg.sql

Reviewer: ```suggestion
4. Create the file `datawarehouse.d/include/Jobs-norm-avg-proc-agg.sql`:
```
